### DISCUSSION TITLE
Should message handlers return current value on `Pmt::Null`?

### DISCUSSION THREAD

**Discussion by metasim** [FutureSDR/FutureSDR#159] [Q&A] [answered]
I've noted in some blocks, such as the Seify `Source` implementation, a message handler for setting a single value will return the current setting upon receiving `Pmt::Null`. For example:

https://github.com/FutureSDR/FutureSDR/blob/fb21d9823f164a9a28e3379b6996613e27cb140f/src/blocks/seify/source.rs#L132

Is this a convention for querying the current setting?

If so, how does one make use of it, given `MessageOutput::post` returns no value?

**Comment by bastibl** [comment]
It's not really a convention. Another alternative would be to have a separate message handler as a getter. I think null is a good approach, but it's not aligned throughout the code base.

**Comment by metasim** [comment]
Followup: How/where does that return value get delivered?

**Comment by bastibl** [comment, marked as answer]
You either call the handler and ignore the return value: https://docs.rs/futuresdr/latest/futuresdr/runtime/struct.FlowgraphHandle.html#method.call
Or you do a callback, which gets you the result: https://docs.rs/futuresdr/latest/futuresdr/runtime/struct.FlowgraphHandle.html#method.callback
In the block, when you post to a message output it is not possible, since there may be no or multiple downstream message blocks connected. So in that case, it wouldn't make sense.
Are you missing some API feature?

**Comment by metasim** [comment]
Thank you. This answers the second question: 

> In the block, when you post to a message output it is not possible, since there may be no or multiple downstream message blocks connected

I was looking for something like a "one-shot" send/receive message capability from one block to another, locally, within the FlowGraph. More of a convenience than a necessity, for inter-block coordination at FlowGraph startup. The point about there being no or many recipients makes sense.

**Reply by metasim** [reply]
May also be I can use `Tag`s for said coordination.

**Reply by bastibl** [reply]
Tags are attached to samples of a stream port. I don't know if this would fit your scenario. It would also not work at flowgraph startup, since streaming only start after all blocks are initialized. Also tags do not provide a back channel.

You could connect message ports between the blocks, i.e., post a message and wait until the other block puts something back in your message handler (this also only works when the flowgraph is running). Or create channels and give the correspondig handles to the blocks in their constructor.
